contributors/devel/api_changes.md

```diff
@@ -146,29 +158,54 @@ type Frobber struct {
 }
 ```
 
-The onus is on you to define a sane default value for `Width` such that rule #1
-above is true - API calls and stored objects that used to work must continue to
-work.
+The onus is on you to define a sane default value for `Width` such that rules
+#1 and #2 above are true - API calls and stored objects that used to work must
+continue to work.
 
 For your next change you want to allow multiple `Param` values. You can not
-simply change `Param string` to `Params []string` (without creating a whole new
-API version) - that fails rules #1 and #2. You can instead do something like:
+simply remove `Param string` and add `Params []string` (without creating a
+whole new API version) - that fails rules #1, #2, #3, and #6.  Nor can you
+simply add `Params []string` and use it instead - that fails #2 and #6.
+
+You must instead define a new field and the relationship between that field and
+the existing field(s).  Start by adding the new plural field:
 
 ```go
-// Still API v6, but kind of clumsy.
+// Still API v6.
 type Frobber struct {
   Height int           `json:"height"`
   Width  int           `json:"width"`
   Param  string        `json:"param"`  // the first param
-  ExtraParams []string `json:"extraParams"` // additional params
+  Params []string      `json:"Params"` // all of the params
 }
 ```
 
-Now you can satisfy the rules: API calls that provide the old style `Param`
-will still work, while servers that don't understand `ExtraParams` can ignore
-it. This is somewhat unsatisfying as an API, but it is strictly compatible.
-
-Part of the reason for versioning APIs and for using internal structs that are
+This new field must be inclusive of the singular field.  In order to satisfy
+the compatibility rules you must handle all the cases of version skew, multiple
+clients, and rollbacks.  This can be handled by defaulting or admission control
+logic linking the fields together with context from the API operation to get as
+close as possible to the user's intentions.
+
+Upon any mutating API operation:
+  * If only the singular field is specified (e.g. an older client), API logic
+    must populate plural[0] from the singular value.
+  * If only the plural field is specified (e.g. a newer client), API logic must
+    populate the singular value from plural[0].
+  * If both the singular and plural fields are specified, API logic must
+    validate that the singular value matches plural[0].
+  * Any other case is an error and must be rejected.
+
+For this purpose "is specified" means the following:
+  * On a create or patch operation: the field is present in the user-provided input
+  * On an update operation: the field is present and has changed from the
+    current value
+
+Older clients that only know the singular field will continue to succeed and
+produce the same results as before the change.  Newer clients can use your
```

Comment:
Probably worth noting that old clients break when trying to update the singular value of plural values are specified (because 0 has changed), since a lot people may not realize the implications

Reply:
Actually you have it below, disregard